### ISSUE TITLE
Suitability report crashes game on OSX

### ISSUE THREAD

**Issue by Vezzra** [freeorion/freeorion#2166]
### Environment

* **FreeOrion Version**: 0.4.8 RC2 ~~(*does not occur on master!*)~~ (maybe on current master too)
* **Operating System**: OSX
* **Fetched as** 
  * Compiled from source

### Description

Only on OSX, ~~and only with builds of the release branch,~~ opening the suitability report for an unowned planet crashes the application (return to desktop), unless you open the suitability report for an owned colony before. If you do the latter first, you can open the suitability report for unowned planets just fine.

Output on the console:
```
main() caught exception(std::exception): Invalid UTF-8
GUIController failed to setup or run the GUI.
```

Logfiles: [Logfiles.zip](https://github.com/freeorion/freeorion/files/2112591/Logfiles.zip)

### Steps to reproduce

* Start a default quickstart game.
* Right click on one of the *unowned* planets in the starting system.
* Application crashes (return to desktop).

**Comment by Vezzra** [comment]
Issue has been reported on the forum: http://freeorion.org/forum/viewtopic.php?f=25&t=10987

**Comment by o01eg** [comment]
> Invalid UTF-8

I suppose it could be error in translation file. Which language is used?

**Comment by Dilvish-fo** [comment]
Vezzra, now that you know how to reliably trigger this bug, could you also double check 0.4.7.1, and then also double check for the issue in a recent-ish build prior to shifting from SDKv8 to SDKv10?  It could be that this is a preexisting error only just now noticed and which has recently been fixed in master serendipitously. 

Comparing the current status of the release branch to master, the changes that seem most likely to account for the different result here also seem related to master being updated to SDKv10.  Specifically related to the UTF-8 error noted above, there are associated changes to StringTable and to i18n.cpp
There are also some changes to various CUI files which seem like the next most likely area to account for the difference.  Those changes appear dependent on the new libraries used in SDKv10.

**Comment by geoffthemedio** [comment]
Where specifically is this exception thrown? ie. can someone attach a debugger and step through the relevant function to find the crash point? Or is that not possible on OSX in release mode?

**Comment by geoffthemedio** [comment]
Might be related to using uncorrected path strings... eg. https://github.com/freeorion/freeorion/blob/master/UI/EncyclopediaDetailPanel.cpp#L2637

Could be instead
`return IsExistingFile(path) && boost::algorithm::starts_with(PathToString(path.filename()), planet_type_str);`

**Comment by Vezzra** [comment]
@geoffthemedio: no, that patch didn't change anything.

> Where specifically is this exception thrown? ie. can someone attach a debugger and step through the relevant function to find the crash point? Or is that not possible on OSX in release mode?

Well, to step through the code with a debugger would require to make a debug build, and even with that I'm not sure if it will work (I tried that a very long time ago, and IIRC these attempts didn't go well). What I can probably get with a debugger is a backtrace. Would that be sufficient?

@Dilvish-fo: I can't reproduce the issue with 0.4.7.1, so I think the bug isn't present there. Next thing I tried was bisecting which commit in master fixed the issue, but that yielded a nonsensical result: 1a23d2e9a42087cc3963968de5234e09dc42a987??? Why on earth would that commit fix the bug?

Didn't have the time to bisect which commit introduced the bug, will try to do that ASAP.

**Comment by geoffthemedio** [comment]
@Vezzra is there anything in the logs from those extra added lines?

You can also add more debug lines into the code generating and the report, which check if the text is valid UTF, similar to what I did above.

A stack trace might help, but might also be in rendering code, which doesn't indicate where the invalid text is being generated.

**Comment by Dilvish-fo** [comment]
>  Next thing I tried was bisecting which commit in master fixed the issue, but that yielded a nonsensical result: 1a23d2e??? Why on earth would that commit fix the bug?

That would be a bit more serendipity than even _my_ optimism allows for 😄

**Comment by Vezzra** [comment]
@geoffthemedio:

> is there anything in the logs from those extra added lines?

Yes:
```
20:49:08.578403 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_asteroids.png  valid utf8?: true
20:49:08.578493 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_barren.png  valid utf8?: true
20:49:08.578527 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_barren02.png  valid utf8?: true
20:49:08.578570 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_desert.png  valid utf8?: true
20:49:08.578600 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_desert02.png  valid utf8?: true
20:49:08.578630 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_gasgiant.png  valid utf8?: true
20:49:08.578658 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_gasgiant02.png  valid utf8?: true
20:49:08.578688 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_inferno.png  valid utf8?: true
20:49:08.578717 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_inferno02.png  valid utf8?: true
20:49:08.578746 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_ocean.png  valid utf8?: true
20:49:08.578775 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_radiated.png  valid utf8?: true
20:49:08.578804 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_swamp.png  valid utf8?: true
20:49:08.578833 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_swamp02.png  valid utf8?: true
20:49:08.578862 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_terran.png  valid utf8?: true
20:49:08.578891 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_toxic.png  valid utf8?: true
20:49:08.578919 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_toxic02.png  valid utf8?: true
20:49:08.578948 [debug] client : EncyclopediaDetailPanel.cpp:2637 : path string: pt_tundra.png  valid utf8?: true
```
Logfiles:  [logfiles.zip](https://github.com/freeorion/freeorion/files/2120672/logfiles.zip)

> You can also add more debug lines into the code generating and the report, which check if the text is valid UTF, similar to what I did above.

I can give it a try, but that probably won't be easy for me as I'm not familiar with the code and won't have a very good idea about what I'm doing. I'll see if I can get to it this weekend.

> A stack trace might help, but might also be in rendering code, which doesn't indicate where the invalid text is being generated.

As it turned out, I can't get a backtrace anyway. I can attach the debugger just fine, the problem is when the crash happens, I get this output in the debugger:
```
Process 9623 exited with status = 0 (0x00000000)
```
When I try to get the backtrace, I get this:
```
(lldb) bt
error: invalid thread
```
I guess because the exception gets caught the process exits cleanly, clearing the backtrace. Bummer ☹️ 

Regarding my attempts to find the commits which introduced and fixed the issue via git bisect, apparently I need to do a complete rebuild each iteration. Builds of the same commit showed different behavior, depending on if I did a complete rebuild or just a "normal" build.

Which means, I have to do all the bisecting again, and this time a clean rebuild with each iteration, which of course is a lot more time consuming (a clean rebuild takes quite some time on my old laptop). This is going to be a tedious process, I hope to be able to do that this weekend too.

Ugh...

**Comment by geoffthemedio** [comment]
hmm. not very helpful if the path is just filenames and not full pathes...

**Comment by Vezzra** [comment]
I've spent many hours these last days trying to narrow down when this bug has been introduced, and unfortunately I'm getting nowhere. That thing is extremely elusive, I finally found out I can't even reliably determine (with reasonable effort) if a certain build has the bug or not (which makes bisecting impossible).

Several restarts of the same build can yield different results, the only thing that apparently varies is the probability of the crash being triggered when invoking the suitability report. Apparently the process of opening the pedia window gets far enough to record in `config.xml` that the pedia window has been opened (even if it doesn't get actually displayed before the app crashes). Which means that restarting after a crash and launching a new game results in the pedia window being open right from start, and it seems as if that alone already influences if the crash occurs or not.

However, even restarting after a crash and launching a game just to close the pedia to ensure you get another "clean" try doesn't work reliably. You can launch a new game twice in a row, get a crash every time (never do an intermediate launch to close the pedia), and the third relaunch will let you open the suitability report without a crash.

Which means I can't even tell for sure if the bug isn't still present in master after all. For all I know the probability for the crash to happen could just be lower for whatever reason. I can't tell when it has been introduced, I can't tell if it still there.

The only thing I can tell for sure is that the earliest commit so far with which I've been able to reproduce the crash is 76284c6ad8205b06089ce4c2d29b452560d4b1c1. As that has been part of the branch for PR #1884, which only touched AI python files, it's probably safe to assume 26474ce5b48072378e8de51abb107cf270c3f78f is already affected too.

I simply don't have the time to track this thing down in reasonable time for 0.4.8. As the issue only turns up under specific conditions, so far only on a platform that now many players are on, and can be worked around quite simply (just open the suitability report for your homeworld at the start of a game, and probably after loading a safe), I'm going to reassing this to optional.

If we get lucky and manage to fix this quickly, we can still get it in, but I don't want to hold up the release for this.

The only thing I think should be done is to note this as a known issue in the changelog, with a short description how to work around it. Any objections, or are you guys ok with that?

Ok, I need a break now, I'm exhausted... 😞

**Comment by geoffthemedio** [comment]
It would probably be more effective to examine the text being set in the pedia window in this function. If it's throwing an invalid UTF8 exception, then it should be possible to narrow down where by adding debug output lines that test that same thing.

**Comment by Vezzra** [comment]
> It would probably be more effective to examine the text being set in the pedia window in this function.

At that point I'm not even sure if the error message we get has anything to do with the actual underlying issue. I mean, if we're really dealing with an invalid UTF-8 string anywhere, why the extremely elusive, unpredictable behavior?

An invalid UTF-8 string should reliably and predictably trigger that exception, not depending on platform, if the pedia/the suitability report had already been (successfully) opened, if the app had already been run several times before, etcpepe.

All that strange behavior points to something very tricky that maybe corrupts the contents of certain memory sections, and just by chance triggers something that comes out as an invalid UTF-8 exception. Which is why I think it's so important to narrow down which set of changes actually introduced the problem.

Of course, seeing how difficult that turned out to be, trying to narrow down where exactly the exception gets thrown is probably the more effective approach at this point. Can't make any promises when I'll find time for this, though.

**Comment by dbenage-cx** [comment]
Using debug logging, the issue appears to be around https://github.com/freeorion/freeorion/blob/master/UI/EncyclopediaDetailPanel.cpp#L2617.  In some instances it does not return from `SingleTabExtent` and in others it does return but stops within the following for loop. (I have yet to catch any instances from `IsValidUTF8()`)

@Vezzra Could you try with this patch? It may not be a complete fix, but I could not reproduce after:
```diff
diff --git a/GG/src/Font.cpp b/GG/src/Font.cpp
index 8142d2e7f..7d53502f0 100644
--- a/GG/src/Font.cpp
+++ b/GG/src/Font.cpp
@@ -1668,6 +1668,11 @@ std::vector<Font::LineData> Font::DetermineLines(
     const std::string& text, Flags<TextFormat>& format, X box_width,
     const std::vector<std::shared_ptr<TextElement>>& text_elements) const
 {
+    if (!utf8::is_valid(text.begin(), text.end())) {
+        std::cout << "Invalid UTF8 in text: " << text;
+        return std::vector<Font::LineData>{};
+    }
+
     ValidateFormat(format);
 
     RenderState render_state;
```

**Comment by Vezzra** [comment]
@dbenage-cx: yep, when I apply this patch to the release branch, I can't reproduce the crash anymore, at least not by the way described in the OP. Without the patch I can. So, at least for the release, this should be good enough.

Can you put up a proper PR? Although the crash does not happen on master, I think it should go into master nevertheless, it's after all a safeguard against a potential error.

What puzzles me, though, is that the condition of the `if` statement is never met (apparently). When I open the suitability report with the patch applied, it displays just fine, and no "Invalid TUF8 in text: <bla>" is printed to the console. Meaning, that change shouldn't have fixed anything.

I'm thoroughly stumped. 😕

**Comment by Vezzra** [comment]
Ok, I'm closing this as resolved for now. Keeping it open will probably just lead to it being forgotten anyway, as it's unlikely anyone will try to work on this as long as the issue doesn't show.

**Comment by Vezzra** [comment]
Addressed by workaround fix #2179.
